source/common/lua/lua.h

```diff
@@ -0,0 +1,378 @@
+
+  virtual ~BaseLuaObject() {}
+
+  /**
+   * Create a new object of this type, owned by Lua. This type must have previously been registered
+   * via the registerType() routine below.
+   * @param state supplies the owning Lua state.
+   * @param args supplies the variadic constructor arguments for the object.
+   * @return a pair containing a pointer to the new object and the state it was created with. (This
+   *         is done for convenience when passing a created object to a LuaRef or a LuaDeathRef.
+   */
+  template <typename... ConstructorArgs>
+  static std::pair<T*, lua_State*> create(lua_State* state, ConstructorArgs&&... args) {
+    // Create a new user data and assign its metatable.
+    void* mem = lua_newuserdata(state, sizeof(T));
+    luaL_getmetatable(state, typeid(T).name());
+    ASSERT(lua_istable(state, -1));
+    lua_setmetatable(state, -2);
+
+    // Memory is allocated via Lua and it is raw. We use placement new to run the constructor.
+    ENVOY_LOG(trace, "creating {} at {}", typeid(T).name(), mem);
+    return {new (mem) T(std::forward<ConstructorArgs>(args)...), state};
+  }
+
+  /**
+   * Register a type with Lua.
+   * @param state supplies the state to register with.
+   */
+  static void registerType(lua_State* state) {
+    std::vector<luaL_Reg> to_register;
+
+    // Fetch all o the functions to be exported to Lua so that we can register them in the
+    // metatable.
+    ExportedFunctions functions = T::exportedFunctions();
+    for (auto function : functions) {
+      to_register.push_back({function.first, function.second});
+    }
+
+    // Always register a __gc method so that we can run the object's destructor. We do this
+    // manually because the memory is raw and was allocated by Lua.
+    to_register.push_back(
+        {"__gc", [](lua_State* state) {
+           T* object = static_cast<T*>(luaL_checkudata(state, 1, typeid(T).name()));
+           ENVOY_LOG(trace, "destroying {} at {}", typeid(T).name(), static_cast<void*>(object));
+           object->~T();
+           return 0;
+         }});
+
+    // Add the sentinel.
+    to_register.push_back({nullptr, nullptr});
+
+    // Register the type by creating a new metatable, setting __index to itself, and then
+    // performing the register.
+    ENVOY_LOG(info, "registering new type: {}", typeid(T).name());
+    int rc = luaL_newmetatable(state, typeid(T).name());
+    ASSERT(rc == 1);
+    UNREFERENCED_PARAMETER(rc);
+
+    lua_pushvalue(state, -1);
+    lua_setfield(state, -2, "__index");
+    luaL_register(state, nullptr, to_register.data());
+  }
+
+  /**
+   * This function is called as part of the DECLARE_LUA_FUNCTION* macros. The idea here is that
+   * we cannot control when Lua destroys things. However, we may expose wrappers to a script that
+   * should not be used after some event. This allows us to mark objects as dead so that if they
+   * are used again they will throw a Lua error and not reach our code.
```

Comment:
I need to read the Lua manual on C binding still, but just from this comment, I'm left wondering how the underlying object has been destroyed by Lua but it still alive enough to generate an error? I.e. if memory is reclaimed, who knows what happens..

Reply:
The point is it hasn't been destroyed by Lua. Inside Lua, an object can a) get GCd whenever, b) can be assigned to a global variable that we have no control over to delay GC. When Lua destroys an object, we will invoke the __gc metamethod and the object is gone. However, if we don't want an object used outside of a certain scope we need to protect ourselves.